project/src/main/ui/chat/career-cutscene-library.gd

```diff
@@ -0,0 +1,309 @@
+var career_cutscene_root_path := DEFAULT_CAREER_CUTSCENE_ROOT_PATH setget set_career_cutscene_root_path
+
+## Ordered list of chat key pairs, from earliest cutscenes to latest cutscenes
+##
+## Each entry in this array is a dictionary with a 'preroll' entry and/or a 'postroll' entry, defining chat keys
+## for cutscenes which play before or after a level.
+var all_chat_key_pairs := [] setget set_all_chat_key_pairs
+
+## key: (String) A preroll chat key like 'chat/career/general_00_a'. For the case where a level has a postroll cutscene
+## 	but no preroll cutscene, this chat key may actually correspond to a non-existent preroll cutscene.
+##
+## value: (Dictionary) A dictionary with a 'preroll' entry and/or a 'postroll' entry, defining chat keys for
+## 	cutscenes which play before or after a level.
+var _chat_key_pairs_by_preroll := {}
+
+## Defines a hierarchy of preroll cutscenes.
+##
+## key: (String) A preroll chat key like 'chat/career/general_00_a', or a parent path like 'chat/career/general'.
+##
+## value: (Array) A list of child chat key fragments, like ['00', '01', '02']. Each fragment corresponds to a suffix
+## 	which can be appended to the chat key, along with an appropriate delimeter, to create a new path.
+var _preroll_tree := {}
+
+func _ready() -> void:
+	_find_all_chat_key_pairs()
+
+
+## Updates the location for career cutscenes. Used for unit tests.
+##
+## Also regenerates all internal fields such as the chat key pairs and preroll tree.
+func set_career_cutscene_root_path(new_career_cutscene_root_path: String) -> void:
+	career_cutscene_root_path = new_career_cutscene_root_path
+	_find_all_chat_key_pairs()
+
+
+## Calculates the list of potential chat key pairs, and returns a random one.
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 	cutscenes to choose between.
+##
+## Returns:
+## 	A dictionary with a 'preroll' entry and/or a 'postroll' entry, defining chat keys for cutscenes which play
+## 	before or after a level.
+func next_chat_key_pair(chat_key_roots: Array) -> Dictionary:
+	var potential_chat_key_pairs := potential_chat_key_pairs(chat_key_roots)
+	return Utils.rand_value(potential_chat_key_pairs) if potential_chat_key_pairs else {}
+
+
+## Calculates the list of potential chat key pairs.
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 	cutscenes to choose between.
+##
+## Returns:
+## 	A list of dictionary entries. Each dictionary entry has a 'preroll' entry and/or a 'postroll' entry, defining
+## 	chat keys for cutscenes which play before or after a level.
+func potential_chat_key_pairs(chat_key_roots: Array) -> Array:
+	var exhausted_chat_keys := _exhausted_chat_keys(chat_key_roots)
+	return _filter_chat_key_pairs(chat_key_roots, exhausted_chat_keys)
+
+
+## Overrides the list of chat key pairs for unit tests.
```

Comment:
This is not just for unit tests.

Reply:
Maybe the field comment could have a blurb like `This is populated based on the contents of the resource folder, but can be overridden for unit tests.`

Reply:
Added blurb for `all_chat_key_pairs` field: `This is calculated based on the contents of the filesystem, but can be overridden for tests.`